File: src/main/java/org/cerion/stocklist/Price.java
package org.cerion.stocklist;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;

public class Price implements Comparable<Price>, IPrice
{
	//Core fields
	public java.util.Date date;
	public float open;
	public float high;
	public float low;
	public float volume;
	private float _close;
	
	//Fields used in list
	public PriceList parent;
	public int pos;

	private static DateFormat mDateFormat = new SimpleDateFormat("yyyy-MM-dd");
	public String getDate() { return mDateFormat.format(date); } //When it needs to be formatted properly
	public static String getDecimal(float val)
	{
		return String.format("%.2f",val);
	}

	public Price(java.util.Date date, float open, float high, float low, float close, float volume) {
		this.date = date;
		this.open = open;
		this.high = high;
		this.low = low;
		this.volume = volume;
		this._close = close;

		//Error checking
		if(open < low || close < low || open > high || close > high)
			throw new RuntimeException("Price range inconsistency " + String.format("%s,%f,%f,%f,%f", getDate(), open, high, low, close));
	}
	
	public int getDOW() {
		Calendar c = Calendar.getInstance();
		c.setTime(date);
		return c.get(Calendar.DAY_OF_WEEK);
	}

	@Deprecated
	public float avgGain() {
		return avgGain(1);
	}

	@Deprecated
	public float avgGain(int days) {
		float gain = 0;
		int end = pos + days;
		
		for(int i = pos+1; i <= end; i++)
		{
			Price t = parent.get(i);
			gain += t.getPercentDiff(this);
		}
		
		gain /= days;	
		
		if(gain > 1000)
			gain = 10; //Bug in PLL
		return gain;
	}

	public float slope(int period) { return parent.slope(period, pos); } //Slope of closing price
	public float tp() { return parent.tp(pos); } //Typical price
	public float change(Price prev)
	{
		return getPercentDiff(prev);
	}
	public float getPercentDiff(Price old)
	{
		if(old.date.before(date) == false)
			throw new RuntimeException("current price is older than input price");
		
		float diff = _close - old._close;
		float percent = (100 * (diff / old._close));
		return percent;
	}

	@Override
	public int compareTo(Price p) {
		//Sort by date ascending, oldest first
		return this.date.compareTo(p.date);
	}

	@Override
	public float getClose() {
		return _close;
	}
}

File: src/main/kotlin/Price.kt
package org.cerion.stocks.core

import java.text.SimpleDateFormat
import java.util.*

class Price(val parent: PriceList, val pos: Int) : IPrice {

    override val date: Date get() = parent.dates[pos]
    override val open: Float get() = parent.open[pos]
    override val close: Float get() = parent.close[pos]
    override val high: Float get() = parent.high[pos]
    override val low: Float get() = parent.low[pos]
    override val volume: Float get() = parent.volume[pos]

    val formattedDate: String
        get() = mDateFormat.format(date) //When it needs to be formatted properly

    val dow: Int
        get() {
            val c = Calendar.getInstance()
            c.time = date
            return c.get(Calendar.DAY_OF_WEEK)
        }

    //Slope of closing price
    fun slope(period: Int): Float {
        return parent.slope(period, pos)
    }

    //Typical price
    fun tp(): Float = parent.tp(pos)
    fun change(prev: Price): Float = getPercentDiff(prev)

    fun getPercentDiff(old: Price): Float {
        if (!old.date.before(date))
            throw RuntimeException("current price is older than input price")

        val diff = close - old.close
        return 100 * (diff / old.close)
    }

    companion object {
        val mDateFormat = SimpleDateFormat("yyyy-MM-dd")

        fun getDecimal(value: Float): String {
            return String.format("%.2f", value)
        }
    }
}
